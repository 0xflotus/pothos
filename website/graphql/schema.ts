import { docsIndex } from '../util/build-index.mjs';
import { builder } from './builder';
import {
  DocsIndex,
  DocsSection,
  SearchMatch,
  SearchMatcher,
  SearchMatchKind,
  SearchResult,
} from './search-matcher';

const docsIndexFiles = docsIndex as DocsIndex[];

const DocsPage = builder.objectRef<DocsIndex>('DocsPage').implement({
  fields: (t) => ({
    title: t.exposeString('name'),
    description: t.exposeString('description'),
    link: t.exposeString('link'),
    md: t.exposeString('md'),
    html: t.exposeString('html'),
  }),
});

builder.enumType(SearchMatchKind, { name: 'SearchMatchKind' });

const SearchMatchRef = builder.objectRef<SearchMatch>('SearchMatch').implement({
  fields: (t) => ({
    kind: t.expose('kind', { type: SearchMatchKind }),
    section: t.expose('section', { type: DocsSectionRef, nullable: true }),
  }),
});

const SearchResultRef = builder.objectRef<SearchResult>('SearchResult').implement({
  fields: (t) => ({
    matches: t.expose('matches', { type: [SearchMatchRef] }),
    doc: t.expose('doc', { type: DocsPage, nullable: true }),
  }),
});

const DocsSectionRef = builder.objectRef<DocsSection>('DocsSection').implement({
  fields: (t) => ({
    heading: t.exposeString('heading'),
    link: t.exposeString('link'),
    md: t.exposeString('md'),
    html: t.exposeString('html'),
  }),
});

builder.queryType({
  fields: (t) => ({
    search: t.field({
      type: [SearchResultRef],
      args: {
        query: t.arg.string({
          required: true,
        }),
      },
      resolve: (_, { query }) => new SearchMatcher(query).matchDocs(docsIndexFiles).getResults(),
    }),
  }),
});

export const schema = builder.toSchema({});
