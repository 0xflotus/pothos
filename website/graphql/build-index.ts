/* eslint-disable unicorn/no-process-exit */
/* eslint-disable no-process-exit */
/* eslint-disable @typescript-eslint/no-implicit-any-catch */
// import algoliasearch from 'algoliasearch/lite';
import dotenv from 'dotenv';

try {
  dotenv.config();

  if (!process.env.NEXT_PUBLIC_ALGOLIA_APP_ID) {
    throw new Error('NEXT_PUBLIC_ALGOLIA_APP_ID is not defined');
  }

  if (!process.env.ALGOLIA_SEARCH_ADMIN_KEY) {
    throw new Error('ALGOLIA_SEARCH_ADMIN_KEY is not defined');
  }
} catch (error) {
  console.error(error);
  process.exit(1);
}

// const searchFormat = loadedFiles.map((file) => {
//   if (!(file.data.title as string)) {
//     throw new Error(`Missing title for ${file.path}`);
//   }

//   return {
//     objectID: file.path,
//     title: file.data.title as string,
//     description: file.data.description as string,
//     slug: (file.data.slug as string) || toSlug(file.path),
//     tagsCollection: { tags: file.data.tags as string[] },
//     date: file.data.date as string,
//     type: 'docs',
//     content: file.content,
//   };
// });
