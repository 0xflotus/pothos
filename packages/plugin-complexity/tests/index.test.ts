import { execute, lexicographicSortSchema, printSchema } from 'graphql';
import { gql } from 'graphql-tag';
import exampleSchema from './example/schema';

describe('simple objects example schema', () => {
  it('generates expected schema', () => {
    expect(printSchema(lexicographicSortSchema(exampleSchema))).toMatchSnapshot();
  });

  describe('queries', () => {
    it('valid query', async () => {
      const query = gql`
        query {
          hero(episode: EMPIRE) {
            friends {
              __typename
              ... on Character {
                name
                friends {
                  friends {
                    name
                  }
                }
              }
            }
          }
        }
      `;

      const result = await execute({
        schema: exampleSchema,
        document: query,
        contextValue: {
          complexity: {
            depth: 5,
            breadth: 10,
            complexity: 200,
          },
        },
      });

      expect(result).toMatchSnapshot();
    });

    it('too complex', async () => {
      const query = gql`
        query {
          hero(episode: EMPIRE) {
            friends {
              ... on Character {
                name
                friends {
                  friends {
                    appearsIn
                    name
                  }
                }
              }
            }
          }
        }
      `;

      const result = await execute({
        schema: exampleSchema,
        document: query,
        contextValue: {
          complexity: {
            depth: 5,
            breadth: 10,
            complexity: 200,
          },
        },
      });

      expect(result).toMatchSnapshot();
    });

    it('complexity based on args', async () => {
      const query = gql`
        query {
          hero(episode: EMPIRE) {
            friends(limit: 1) {
              ... on Character {
                name
                friends(limit: 1) {
                  friends(limit: 1) {
                    appearsIn
                    name
                  }
                }
              }
            }
          }
        }
      `;

      const result = await execute({
        schema: exampleSchema,
        document: query,
        contextValue: {
          complexity: {
            depth: 5,
            breadth: 10,
            complexity: 200,
          },
        },
      });

      expect(result).toMatchSnapshot();
    });

    it('complexity based options', async () => {
      const query = gql`
        query {
          human(id: 1) {
            name
          }
        }
      `;

      const result = await execute({
        schema: exampleSchema,
        document: query,
        contextValue: {
          complexity: {
            depth: 5,
            breadth: 10,
            complexity: 200,
          },
        },
      });

      expect(result).toMatchSnapshot();
    });

    it('too deep', async () => {
      const query = gql`
        query {
          hero(episode: EMPIRE) {
            friends {
              ... on Character {
                name
                friends {
                  friends {
                    friends {
                      name
                    }
                  }
                }
              }
            }
          }
        }
      `;

      const result = await execute({
        schema: exampleSchema,
        document: query,
        contextValue: {
          complexity: {
            depth: 5,
            breadth: 10,
            complexity: 200,
          },
        },
      });

      expect(result).toMatchSnapshot();
    });

    it('too wide', async () => {
      const query = gql`
        query {
          hero(episode: EMPIRE) {
            name1: name
            name2: name
            name3: name
            name4: name
            ... {
              name1: name
              name2: name
              name3: name
              name4: name
            }
            ... on Character {
              name1: name
              name2: name
              name3: name
              name4: name
            }
          }
        }
      `;

      const result = await execute({
        schema: exampleSchema,
        document: query,
        contextValue: {
          complexity: {
            depth: 5,
            breadth: 10,
            complexity: 200,
          },
        },
      });

      expect(result).toMatchSnapshot();
    });
  });
});
