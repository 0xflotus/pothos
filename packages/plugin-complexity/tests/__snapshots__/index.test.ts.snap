// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple objects example schema generates expected schema 1`] = `
"interface Character {
  \\"\\"\\"Which movies they appear in.\\"\\"\\"
  appearsIn: [Episode!]!

  \\"\\"\\"The friends of the character, or an empty list if they have none.\\"\\"\\"
  friends(limit: Int): [Character]!

  \\"\\"\\"The id of the character\\"\\"\\"
  id: ID!

  \\"\\"\\"The name of the character\\"\\"\\"
  name: String!
}

\\"\\"\\"A mechanical creature in the Star Wars universe.\\"\\"\\"
type Droid implements Character {
  \\"\\"\\"Which movies they appear in.\\"\\"\\"
  appearsIn: [Episode!]!

  \\"\\"\\"The friends of the character, or an empty list if they have none.\\"\\"\\"
  friends(limit: Int): [Character]!

  \\"\\"\\"The id of the character\\"\\"\\"
  id: ID!

  \\"\\"\\"The name of the character\\"\\"\\"
  name: String!

  \\"\\"\\"The primary function of the droid.\\"\\"\\"
  primaryFunction: String!
}

\\"\\"\\"One of the films in the Star Wars Trilogy\\"\\"\\"
enum Episode {
  \\"\\"\\"Released in 1980.\\"\\"\\"
  EMPIRE

  \\"\\"\\"Released in 1983\\"\\"\\"
  JEDI

  \\"\\"\\"Released in 1977.\\"\\"\\"
  NEWHOPE
}

\\"\\"\\"A humanoid creature in the Star Wars universe.\\"\\"\\"
type Human implements Character {
  \\"\\"\\"Which movies they appear in.\\"\\"\\"
  appearsIn: [Episode!]!

  \\"\\"\\"The friends of the character, or an empty list if they have none.\\"\\"\\"
  friends(limit: Int): [Character]!

  \\"\\"\\"The home planet of the human, or null if unknown.\\"\\"\\"
  homePlanet: String

  \\"\\"\\"The id of the character\\"\\"\\"
  id: ID!

  \\"\\"\\"The name of the character\\"\\"\\"
  name: String!
}

type Query {
  droid(
    \\"\\"\\"id of the character\\"\\"\\"
    id: ID!
  ): Droid!
  hero(
    \\"\\"\\"
    If omitted, returns the hero of the whole saga. If provided, returns the hero of that particular episode.
    \\"\\"\\"
    episode: Episode!
  ): Character!
  human(
    \\"\\"\\"id of the character\\"\\"\\"
    id: ID!
  ): Human!
  r2d2: Droid!
}"
`;

exports[`simple objects example schema queries complexity based on args 1`] = `
Object {
  "data": Object {
    "hero": Object {
      "friends": Array [
        Object {
          "friends": Array [
            Object {
              "friends": Array [
                Object {
                  "appearsIn": Array [
                    "NEWHOPE",
                    "EMPIRE",
                    "JEDI",
                  ],
                  "name": "Han Solo",
                },
              ],
            },
          ],
          "name": "Han Solo",
        },
      ],
    },
  },
}
`;

exports[`simple objects example schema queries complexity based options 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Query exceeds maximum complexity (complexity: 501, max: 200)],
  ],
}
`;

exports[`simple objects example schema queries too complex 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Query exceeds maximum complexity (complexity: 287, max: 200)],
  ],
}
`;

exports[`simple objects example schema queries too deep 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Query exceeds maximum depth (depth: 6, max: 5)],
  ],
}
`;

exports[`simple objects example schema queries too wide 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Query exceeds maximum breadth (breadth: 13, max: 10)],
  ],
}
`;

exports[`simple objects example schema queries valid query 1`] = `
Object {
  "data": Object {
    "hero": Object {
      "friends": Array [
        Object {
          "__typename": "Human",
          "friends": Array [
            Object {
              "friends": Array [
                Object {
                  "name": "Han Solo",
                },
                Object {
                  "name": "Leia Organa",
                },
                Object {
                  "name": "C-3PO",
                },
                Object {
                  "name": "R2-D2",
                },
              ],
            },
            Object {
              "friends": Array [
                Object {
                  "name": "Luke Skywalker",
                },
                Object {
                  "name": "Han Solo",
                },
                Object {
                  "name": "C-3PO",
                },
                Object {
                  "name": "R2-D2",
                },
              ],
            },
            Object {
              "friends": Array [
                Object {
                  "name": "Luke Skywalker",
                },
                Object {
                  "name": "Han Solo",
                },
                Object {
                  "name": "Leia Organa",
                },
              ],
            },
          ],
          "name": "Han Solo",
        },
        Object {
          "__typename": "Human",
          "friends": Array [
            Object {
              "friends": Array [
                Object {
                  "name": "Han Solo",
                },
                Object {
                  "name": "Leia Organa",
                },
                Object {
                  "name": "C-3PO",
                },
                Object {
                  "name": "R2-D2",
                },
              ],
            },
            Object {
              "friends": Array [
                Object {
                  "name": "Luke Skywalker",
                },
                Object {
                  "name": "Leia Organa",
                },
                Object {
                  "name": "R2-D2",
                },
              ],
            },
            Object {
              "friends": Array [
                Object {
                  "name": "Luke Skywalker",
                },
                Object {
                  "name": "Han Solo",
                },
                Object {
                  "name": "Leia Organa",
                },
                Object {
                  "name": "R2-D2",
                },
              ],
            },
            Object {
              "friends": Array [
                Object {
                  "name": "Luke Skywalker",
                },
                Object {
                  "name": "Han Solo",
                },
                Object {
                  "name": "Leia Organa",
                },
              ],
            },
          ],
          "name": "Leia Organa",
        },
        Object {
          "__typename": "Droid",
          "friends": Array [
            Object {
              "friends": Array [
                Object {
                  "name": "Han Solo",
                },
                Object {
                  "name": "Leia Organa",
                },
                Object {
                  "name": "C-3PO",
                },
                Object {
                  "name": "R2-D2",
                },
              ],
            },
            Object {
              "friends": Array [
                Object {
                  "name": "Luke Skywalker",
                },
                Object {
                  "name": "Leia Organa",
                },
                Object {
                  "name": "R2-D2",
                },
              ],
            },
            Object {
              "friends": Array [
                Object {
                  "name": "Luke Skywalker",
                },
                Object {
                  "name": "Han Solo",
                },
                Object {
                  "name": "C-3PO",
                },
                Object {
                  "name": "R2-D2",
                },
              ],
            },
            Object {
              "friends": Array [
                Object {
                  "name": "Luke Skywalker",
                },
                Object {
                  "name": "Han Solo",
                },
                Object {
                  "name": "Leia Organa",
                },
              ],
            },
          ],
          "name": "C-3PO",
        },
        Object {
          "__typename": "Droid",
          "friends": Array [
            Object {
              "friends": Array [
                Object {
                  "name": "Han Solo",
                },
                Object {
                  "name": "Leia Organa",
                },
                Object {
                  "name": "C-3PO",
                },
                Object {
                  "name": "R2-D2",
                },
              ],
            },
            Object {
              "friends": Array [
                Object {
                  "name": "Luke Skywalker",
                },
                Object {
                  "name": "Leia Organa",
                },
                Object {
                  "name": "R2-D2",
                },
              ],
            },
            Object {
              "friends": Array [
                Object {
                  "name": "Luke Skywalker",
                },
                Object {
                  "name": "Han Solo",
                },
                Object {
                  "name": "C-3PO",
                },
                Object {
                  "name": "R2-D2",
                },
              ],
            },
          ],
          "name": "R2-D2",
        },
      ],
    },
  },
}
`;
